python/modules/IcePy/Util.cpp

```diff
@@ -291,204 +291,171 @@ IcePy::PyException::PyException(PyObject* raisedException)
     _tb = PyException_GetTraceback(raisedException);
 }
 
-void
-IcePy::PyException::raise()
+namespace IcePy
 {
-    assert(ex.get());
-
-    PyObject* userExceptionType = lookupType("Ice.UserException");
-    PyObject* localExceptionType = lookupType("Ice.LocalException");
-
-    // TODO: create better error messages.
+    string getTypeName(PyObject* ex)
+    {
+        PyObject* cls = reinterpret_cast<PyObject*>(ex->ob_type);
+        PyObjectHandle name = getAttr(cls, "__name__", false);
+        assert(name.get());
+        PyObjectHandle mod = getAttr(cls, "__module__", false);
+        assert(mod.get());
+        string result = getString(mod.get());
+        result += ".";
+        result += getString(name.get());
+        return result;
+    }
 
-    if (PyObject_IsInstance(ex.get(), userExceptionType))
+    string getTraceback(PyObject* type, PyObject* ex, PyObject* tb)
     {
-        string tb = getTraceback();
-        if (!tb.empty())
-        {
-            throw Ice::UnknownUserException{__FILE__, __LINE__, tb};
-        }
-        else
+        ostringstream os;
+        assert(tb);
+        //
+        // We need the equivalent of the following Python code:
+        //
+        // import traceback
+        // list = traceback.format_exception(type, ex, tb)
+        //
+        PyObjectHandle str = createString("traceback");
+        PyObjectHandle mod = PyImport_Import(str.get());
+        assert(mod.get()); // Unable to import traceback module - Python installation error?
+        PyObject* func = PyDict_GetItemString(PyModule_GetDict(mod.get()), "format_exception");
+        assert(func); // traceback.format_exception must be present.
+        PyObjectHandle args = Py_BuildValue("(OOO)", type, ex, tb);
+        assert(args.get());
+        PyObjectHandle list = PyObject_CallObject(func, args.get());
+        assert(list.get());
+
+        for (Py_ssize_t i = 0; i < PyList_GET_SIZE(list.get()); ++i)
         {
-            PyObjectHandle name = PyObject_CallMethod(ex.get(), "ice_id", 0);
-            PyErr_Clear();
-            if (!name.get())
-            {
-                throw Ice::UnknownUserException{__FILE__, __LINE__, getTypeName()};
-            }
-            else
-            {
-                throw Ice::UnknownUserException{__FILE__, __LINE__, getString(name.get())};
-            }
+            os << getString(PyList_GetItem(list.get(), i));
         }
+        return os.str();
     }
-    else if (PyObject_IsInstance(ex.get(), localExceptionType))
-    {
-        raiseLocalException();
-    }
-    else
+
+    string createUnknownExceptionMessage(PyObject* ex, PyObject* type, PyObject* tb, bool includeStackTrace)
```

Comment:
I don't understand why we have an includeStackTrace parameter here (and by extension, the type and tb parameters).

The message for an UnknownException never includes the stack trace.

Reply:
The intention was to include it when it is raised by the application, and not include it when it is raised by the remote peer.